src/sage/data_structures/pairing_heap.h

```diff
@@ -0,0 +1,370 @@
+ *
+ * - contains(item): check whether specified item is in the heap in time O(1).
+ *
+ * - value(item): return the value associated with the item in the heap.
+ *   This operation assumes that the item is already in the heap.
+ *
+ * References:
+ *
+ * [1] M. L. Fredman, R. Sedgewick, D. D. Sleator, and R. E. Tarjan.
+ *     "The pairing heap: a new form of self-adjusting heap".
+ *     Algorithmica. 1 (1): 111-129, 1986. doi:10.1007/BF01840439.
+ *
+ * [2] https://en.wikipedia.org/wiki/Pairing_heap
+ *
+ * Author:
+ * - David Coudert <david.coudert@inria.fr>
+ *
+ */
+
+#ifndef PAIRING_HEAP_H
+#define PAIRING_HEAP_H
+
+#include <iostream>
+#include <unordered_map>
+#include <stdexcept>
+
+
+namespace pairing_heap {
+
+  template<
+    typename TI,  // type of items stored in the node
+    typename TV   // type of values associated with the stored item
+    >
+    struct PairingHeapNode {
+      TI item;   // item contained in the node
+      TV value;  // value associated with the item
+
+      PairingHeapNode<TI, TV> * prev;  // Previous sibling of the node or parent
+      PairingHeapNode<TI, TV> * next;  // Next sibling of the node
+      PairingHeapNode<TI, TV> * child; // First child of the node
+
+      explicit PairingHeapNode(const TI &some_item, const TV &some_value):
+      item(some_item), value(some_value), prev(nullptr), next(nullptr), child(nullptr) {
+      }
+
+      bool operator<(PairingHeapNode const& other) const {
+	return value < other.value;
+      }
+
+      bool operator<=(PairingHeapNode const& other) const {
+	return value <= other.value;
+      }
+    };
+
+
+  template<
+    typename TI,  // type of items stored in the node
+    typename TV   // type of values associated with the stored item
+    >
+    class PairingHeap
+    {
+    public:
+      // Constructor
+      explicit PairingHeap();
+
+      // Copy constructor
+      PairingHeap(PairingHeap<TI, TV> const *other);
+
+      // Destructor
+      virtual ~PairingHeap();
+
+      // Return true if the heap is empty, else false
+      bool empty()  { return root == nullptr; }
+
+      // Return true if the heap is empty, else false
+      explicit operator bool() const { return root != nullptr; }
+
+      // Insert an item into the heap with specified value (priority)
+      void push(const TI &some_item, const TV &some_value);
+
+      // Return the top pair (item, value) of the heap
+      std::pair<TI, TV> top();
+
+      // Return the top item of the heap
+      TI top_item();
+
+      // Return the top value of the heap
+      TV top_value();
+
+      // Remove the top element from the heap
+      void pop();
+
+      // Decrease the value of specified item
+      void decrease(const TI &some_item, const TV &new_value);
+
+      // Check if specified item is in the heap
+      bool contains(TI const& some_item);
+
+      // Return the value associated with the item
+      TV value(const TI &some_item);
```

Comment:
When I try that, I get compilation errors...

Reply:
I see the problem: in the implementation you use `nodes[some_item]`:
```c++
      if (nodes.find(some_item) == nodes.end()) {
	throw std::invalid_argument("the specified item is not in the heap");
      }
      return nodes[some_item]->value;
```
But the `operator[]` creates the element in the map if it does not exist, so it can modify `nodes`, so the method cannot be `const` (the compiler here is not smart enough to see that the check above ensures that `nodes` would not be modified)
You can circumvent this problem by using the iterator return by `find`:
```c++
     auto it = nodes.find(some_item);
      if (it == nodes.end()) {
	throw std::invalid_argument("the specified item is not in the heap");
      }
      return it->value;
```

Reply:
even without the const, I get another issue :(
```
[sagelib-10.5.rc1] [spkg-install]     In file included from sage/data_structures/pairing_heap.cpp:1275:
[sagelib-10.5.rc1] [spkg-install]     sage/data_structures/./pairing_heap.h:268:18: error: no member named 'value' in 'std::pair<const unsigned long, pairing_heap::PairingHeapNode<unsigned long, unsigned long> *>'
[sagelib-10.5.rc1] [spkg-install]       268 |       return it->value;
[sagelib-10.5.rc1] [spkg-install]           |              ~~~~^
[sagelib-10.5.rc1] [spkg-install]     sage/data_structures/pairing_heap.cpp:19820:32: note: in instantiation of member function 'pairing_heap::PairingHeap<unsigned long, unsigned long>::value' requested here
[sagelib-10.5.rc1] [spkg-install]      19820 |       __pyx_t_17 = __pyx_v_PH->value(__pyx_t_12);
[sagelib-10.5.rc1] [spkg-install]            |                                ^
[sagelib-10.5.rc1] [spkg-install]     In file included from sage/data_structures/pairing_heap.cpp:1275:
[sagelib-10.5.rc1] [spkg-install]     sage/data_structures/./pairing_heap.h:268:18: error: no member named 'value' in 'std::pair<const unsigned long, pairing_heap::PairingHeapNode<unsigned long, std::pair<unsigned long, unsigned long>> *>'
[sagelib-10.5.rc1] [spkg-install]       268 |       return it->value;
[sagelib-10.5.rc1] [spkg-install]           |              ~~~~^
[sagelib-10.5.rc1] [spkg-install]     sage/data_structures/pairing_heap.cpp:20961:34: note: in instantiation of member function 'pairing_heap::PairingHeap<unsigned long, std::pair<unsigned long, unsigned long>>::value' requested here
[sagelib-10.5.rc1] [spkg-install]      20961 |         __pyx_t_20 = __pyx_v_HH->value(0x7B);
[sagelib-10.5.rc1] [spkg-install]            |                                  ^
[sagelib-10.5.rc1] [spkg-install]     2 errors generated.
[sagelib-10.5.rc1] [spkg-install]     error: command '/usr/bin/g++' failed with exit code 1
[sagelib-10.5.rc1] [spkg-install]     error: subprocess-exited-with-error
```

Reply:
the right form is `return it->second->value;`